XMonad/Actions/MostRecentlyUsed.hs

```diff
@@ -0,0 +1,179 @@
+import Data.List.NonEmpty (NonEmpty(..), nonEmpty)
+import Data.Monoid (All(..))
+import Data.Foldable (for_)
+import Data.Functor (($>))
+import Control.Monad (when, unless)
+
+-- mtl
+import Control.Monad.Trans (lift)
+import Control.Monad.State (get, put, gets)
+
+-- containers
+import qualified Data.Map.Strict as M
+
+-- xmonad
+import XMonad
+  ( Window, KeySym, keyPress
+  , Event (DestroyWindowEvent, UnmapEvent, ev_send_event, ev_window)
+  )
+import XMonad.Core
+  ( X, XConfig(..), windowset, WorkspaceId
+  , ExtensionClass(..), StateExtension(..)
+  , waitingUnmap
+  )
+import XMonad.Operations (windows)
+import qualified XMonad.StackSet as W
+
+-- xmonad-contrib
+import qualified XMonad.Util.ExtensibleConf  as XC
+import qualified XMonad.Util.ExtensibleState as XS
+import XMonad.Util.History (History, origin, event, erase, ledger)
+import XMonad.Actions.Repeatable (repeatableSt)
+
+-- }}}
+
+-- --< Core Data Type: WindowHistory >-- {{{
+
+data WindowHistory = WinHist
+  { busy :: !Bool
+  , hist :: !(History Window WorkspaceId)
+  } deriving (Show, Read)
+
+instance ExtensionClass WindowHistory where
+  initialValue = WinHist
+    { busy = False
+    , hist = origin
+    }
+  extensionType = PersistentExtension
+
+-- }}}
+
+-- --< Interface >-- {{{
+
+-- $usage
+--
+-- 'configureMRU' must be applied to your config in order for 'mostRecentlyUsed'
+-- to work.
+--
+-- > main :: IO ()
+-- > main = xmonad . configureMRU . ... $ def
+-- >   { ...
+-- >   }
+--
+-- Once that's done, it can be used normally in keybinds:
+--
+-- > , ((mod1Mask, xK_Tab), mostRecentlyUsed [xK_Alt_L, xK_Alt_R] xK_Tab)
+--
+-- N.B.: This example assumes that 'mod1Mask' corresponds to alt, which is not
+-- always the case, depending on how your system is configured.
+
+-- | Configure xmonad to support 'mostRecentlyUsed'.
+configureMRU :: XConfig l -> XConfig l
+configureMRU = XC.once f (MRU ()) where
+  f cnf = cnf
+    { logHook         = logHook         cnf <> logWinHist
+    , handleEventHook = handleEventHook cnf <> winHistEH
+    }
+newtype MRU = MRU () deriving Semigroup
+
+-- | An action to browse through the history of focused windows, taking
+--   another step back with each tap of the key.
+mostRecentlyUsed
+  :: [KeySym] -- ^ The 'KeySym's corresponding to the modifier to which the
+              --   action is bound.
+  -> KeySym   -- ^ The 'KeySym' corresponding to the key to which the action
+              --   is bound.
+  -> X ()
+mostRecentlyUsed mods tab = do
+  wh@WinHist{busy,hist} <- XS.get
+  unless busy $ do
+    XS.put wh{ busy = True }
+
+    for_ (nonEmpty $ ledger hist) $ \ne -> do
+      mfw <- gets (W.peek . windowset)
+      let iSt = case cycleS ne of
+            (w, _) :~ s | mfw == Just w -> s
+            s                           -> s
+      repeatableSt iSt mods $ \t s ->
+        when (t == keyPress && s == tab) $ do
+          (win, wksp) <- pop
+          lift . windows $ W.focusWindow win . W.greedyView wksp
```

Comment:
There is an initial "delay" of one key press here when one first starts cycling.  Suppose we have

``` haskell
    ("M-<Tab>", mostRecentlyUsed [xK_Super_L, xK_Super_R] xK_Tab)
```

When `M-<Tab>` is pressed, we essentially come to this event loop and stop, listening for another keypress (e.g. another `M-<Tab>`, to start cycling).  However, intuitively the first `M-<Tab>` should already start the cycling, not requiring another confirmation.

Reply:
You're right; all the other repeatable actions do some special handling for the first keypress ... but this is actually a weakness in `repeatableAction`, which I cribbed without thinking too hard about. Rather than opening with `getNextEvent >>= handleEvent`, it should do `handleEvent (keyPress, key)`, then the special handling can be erased.

This requires passing an additional argument to the `repeatable*` functions, but it's a value that's on hand, since it's needed to write the keypress handler anyway. The only problem is in `repeatableAction` itself, which now requires a change of type to be written in terms of `repeatable`. I wonder if it's worth temporarily undeleting the original implementation?

Reply:
Tbh, I consider `repeatableAction` to be pretty "internal": I doubt many users are actually using this and not just a wrapper that'll remain unaffected. 

@LSLeary this is pretty much done, right?  No further complains from my side, so whenever you're ready you can rebase and merge this